File: packages/raisin-client-java/examples/BasicUsage.java
import com.raisindb.client.RaisinClient;
import com.raisindb.client.operations.*;
import com.raisindb.client.protocol.*;
import com.raisindb.client.events.Subscription;

import java.util.*;

/**
 * Basic usage example for RaisinDB Java client.
 */
public class BasicUsage {

    public static void main(String[] args) {
        try (RaisinClient client = new RaisinClient("raisin://localhost:8080/sys/default")) {

            // Connect to RaisinDB
            client.connect();
            System.out.println("✓ Connected to RaisinDB");

            // Authenticate
            client.authenticate("admin", "admin").get();
            System.out.println("✓ Authenticated");

            // Get database and workspace
            Database db = client.database("demo");
            Workspace workspace = db.workspace("content");

            // Create a node
            System.out.println("\nCreating a node...");
            Map<String, Object> properties = new HashMap<>();
            properties.put("title", "Welcome Page");
            properties.put("published", true);
            properties.put("author", "admin");

            Node node = workspace.nodes().create(
                    "Page",
                    "/welcome",
                    properties,
                    Map.of("body", "Welcome to RaisinDB!")
            ).get();

            System.out.println("✓ Created node: " + node.getNodeId());
            System.out.println("  Path: " + node.getPath());
            System.out.println("  Properties: " + node.getProperties());

            // Query nodes by path
            System.out.println("\nQuerying nodes...");
            List<Node> nodes = workspace.nodes().queryByPath("/welcome").get();
            System.out.println("✓ Found " + nodes.size() + " node(s)");
            for (Node n : nodes) {
                System.out.println("  - " + n.getPath() + ": " +
                        n.getProperties().get("title"));
            }

            // Update the node
            System.out.println("\nUpdating node...");
            Map<String, Object> updates = new HashMap<>();
            updates.put("title", "Updated Welcome Page");
            updates.put("published", true);

            Node updatedNode = workspace.nodes().update(
                    node.getNodeId(),
                    updates,
                    null
            ).get();
            System.out.println("✓ Updated node: " + updatedNode.getNodeId());
            System.out.println("  New title: " + updatedNode.getProperties().get("title"));

            // Execute SQL query
            System.out.println("\nExecuting SQL query...");
            SqlResult result = db.sql(
                    "SELECT * FROM nodes WHERE node_type = ?",
                    "Page"
            ).get();
            System.out.println("✓ Query returned " + result.getRowCount() + " row(s)");
            System.out.println("  Columns: " + result.getColumns());

            // List workspaces
            System.out.println("\nListing workspaces...");
            List<Object> workspaces = db.listWorkspaces().get();
            System.out.println("✓ Found " + workspaces.size() + " workspace(s)");

            // Subscribe to events
            System.out.println("\nSubscribing to events...");
            final int[] eventCount = {0};

            Subscription subscription = workspace.events().subscribe(
                    event -> {
                        eventCount[0]++;
                        System.out.println("\n📢 Event received: " + event.getEventType());
                        System.out.println("   Subscription: " + event.getSubscriptionId());
                        System.out.println("   Timestamp: " + event.getTimestamp());
                    },
                    "/",
                    List.of("node:created", "node:updated"),
                    null
            ).get();
            System.out.println("✓ Subscribed with ID: " + subscription.getSubscriptionId());

            // Create another node to trigger event
            System.out.println("\nCreating another node (should trigger event)...");
            Node node2 = workspace.nodes().create(
                    "Page",
                    "/test",
                    Map.of("title", "Test Page"),
                    null
            ).get();
            System.out.println("✓ Created node: " + node2.getNodeId());

            // Wait a bit for events
            Thread.sleep(1000);

            // Unsubscribe
            subscription.unsubscribe().get();
            System.out.println("✓ Unsubscribed (received " + eventCount[0] + " event(s))");

            // Clean up - delete test nodes
            System.out.println("\nCleaning up...");
            workspace.nodes().delete(node.getNodeId()).get();
            workspace.nodes().delete(node2.getNodeId()).get();
            System.out.println("✓ Deleted test nodes");

            System.out.println("\n✓ Example completed successfully!");

        } catch (Exception e) {
            System.err.println("✗ Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: packages/raisin-client-py/examples/basic_usage.py
"""Basic usage example for RaisinDB Python client."""

import asyncio
import logging

from raisindb import RaisinClient

# Enable logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)


async def main() -> None:
    """Main example function."""
    # Create and connect client
    client = RaisinClient("raisin://localhost:8080/sys/default")

    try:
        await client.connect()
        print("✓ Connected to RaisinDB")

        # Authenticate
        await client.authenticate("admin", "admin")
        print("✓ Authenticated")

        # Get database and workspace
        db = client.database("demo")
        workspace = db.workspace("content")

        # Create a node
        print("\nCreating a node...")
        node = await workspace.nodes().create(
            node_type="Page",
            path="/welcome",
            properties={"title": "Welcome Page", "published": True, "author": "admin"},
            content={"body": "Welcome to RaisinDB!"},
        )
        print(f"✓ Created node: {node.node_id}")
        print(f"  Path: {node.path}")
        print(f"  Properties: {node.properties}")

        # Query nodes by path
        print("\nQuerying nodes...")
        nodes = await workspace.nodes().query_by_path("/welcome")
        print(f"✓ Found {len(nodes)} node(s)")
        for n in nodes:
            print(f"  - {n.path}: {n.properties.get('title')}")

        # Update the node
        print("\nUpdating node...")
        updated_node = await workspace.nodes().update(
            node_id=node.node_id,
            properties={"title": "Updated Welcome Page", "published": True},
        )
        print(f"✓ Updated node: {updated_node.node_id}")
        print(f"  New title: {updated_node.properties.get('title')}")

        # Execute SQL query
        print("\nExecuting SQL query...")
        result = await db.sql("SELECT * FROM nodes WHERE node_type = ?", "Page")
        print(f"✓ Query returned {result.row_count} row(s)")
        print(f"  Columns: {result.columns}")
        for row in result.rows:
            print(f"  Row: {row}")

        # List workspaces
        print("\nListing workspaces...")
        workspaces = await db.list_workspaces()
        print(f"✓ Found {len(workspaces)} workspace(s)")
        for ws in workspaces:
            print(f"  - {ws}")

        # Subscribe to events
        print("\nSubscribing to events...")

        event_count = 0

        async def on_event(event):
            nonlocal event_count
            event_count += 1
            print(f"\n📢 Event received: {event.event_type}")
            print(f"   Subscription: {event.subscription_id}")
            print(f"   Timestamp: {event.timestamp}")
            print(f"   Payload: {event.payload}")

        subscription = await workspace.events().subscribe(
            callback=on_event, path="/", event_types=["node:created", "node:updated"]
        )
        print(f"✓ Subscribed with ID: {subscription.subscription_id}")

        # Create another node to trigger event
        print("\nCreating another node (should trigger event)...")
        node2 = await workspace.nodes().create(
            node_type="Page",
            path="/test",
            properties={"title": "Test Page"},
        )
        print(f"✓ Created node: {node2.node_id}")

        # Wait a bit for events
        await asyncio.sleep(1)

        # Unsubscribe
        await subscription.unsubscribe()
        print(f"✓ Unsubscribed (received {event_count} event(s))")

        # Clean up - delete test nodes
        print("\nCleaning up...")
        await workspace.nodes().delete(node.node_id)
        await workspace.nodes().delete(node2.node_id)
        print("✓ Deleted test nodes")

    except Exception as e:
        print(f"✗ Error: {e}")
        raise

    finally:
        # Close connection
        await client.close()
        print("\n✓ Connection closed")


if __name__ == "__main__":
    asyncio.run(main())
